### PR TITLE
[WIP] OA failsafe

### PR DESCRIPTION
**Pull request by baumanta**
We need a way to trigger a failsafe in the case where the obstacle avoidance dies or does not provide input for any other reason.

Problem:
We are in Mission mode and the obstacle avoidance is enabled (MPC_OBS_AVOID = true). The planner provides setpoints to guide the drone around any obstacles which lie in the straight path between the mission points. If the planner dies, the regular mission is resumed without the planner input and the drone just flies straight to the next mission point. This can be very dangerous if it happens close to obstacles. 

Solution:
If in Mission mode and OA is enabled, check if setpoints from planner are received. If there is a timeout on these points, take failsafe action.

Note: I'm not sure whether this is an acceptable way to solve this problem, or if the logic for that should go into commander. I'd be glad for any input on solving this issue.

### CODE REVIEW

**Comment by hamishwillee** [comment]
FMI, what are the failsafe options you plan on offering? 
Common options are: https://docs.px4.io/en/config/safety.html#failsafe_actions

**Comment by baumanta** [comment]
@hamishwillee I'm not very familiar with the failsafe options and how to trigger them individually. What I did in the code was switch to the failsafe flight task. I thought it would choose and appropriate reaction automatically. When I tested this code in SITL the drone went into some kind of LOITER/HOLD state when I killed the planner. This would probably be the preferred reaction. To me it is most important that the drone aborts the mission and does not just continue on straight line (as this caused a crash lately).

**Comment by hamishwillee** [comment]
> @hamishwillee I'm not very familiar with the failsafe options and how to trigger them individually. 

I'm not either. If I had to do this I would ask @bkueng what he thinks. Beat, that's what you get for knowing things :-)

> What I did in the code was switch to the failsafe flight task. I thought it would choose and appropriate reaction automatically. When I tested this code in SITL the drone went into some kind of LOITER/HOLD state when I killed the planner. This would probably be the preferred reaction. To me it is most important that the drone aborts the mission and does not just continue on straight line (as this caused a crash lately).

FWIW IMO Hold mode is a fine default option because there is a good chance that with obstacle avoidance people will plan more risky missions "altitude-wise" than otherwise - so proceeding with the normal mission is dangerous. But I do think that this is something that you _might_ want to offer as an option. As a user I _would_ like the option to instead choose RTL - that is likely to be safe in most cases because it starts with ascending to a safe level.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -70,6 +70,12 @@
 
 using namespace time_literals;
 
+enum class ObstacleAvoidanceStatus {
```
[the remark this answers is not in the document]
Comment (baumanta) [reply src/modules/mc_pos_control/mc_pos_control_main.cpp]: there is no such thing as an obstacle avoidance library until now. There is the collision prevention library which is something slightly different. The collision prevention works for position control (manual modes) only. The whole logic is done in the firmware. The obstacle avoidance, on the other hand, runs on the companion computer and modifies the /trajectory message which is used in mission mode. The whole interface for updating this trajectory message is in the position controller

**Comment by MaEtUgR** [comment]
I checked and apparently the failsafe when the `update()` function returns false is different which should not be the case. Let me find a fix for that. Then we can just trigger it.

**Comment by MaEtUgR** [comment]
@baumanta Fixed by #11187 ? We can still improve on it.

**Pull request closed without merging**